### DISCUSSION TITLE
Omni Mode is not handled according to MIDI spec

### DISCUSSION THREAD

**Discussion by rianhunter** [FluidSynth/fluidsynth#1186] [General]
According to the MIDI spec when the device is Omni mode, the basic channel is ignored:

> The recommended start up condition is Omni Mode On. This allows two instruments to be connected and played immediately without concern for selecting the instruments' basic channel.

In FluidSynth, even when in omni mode, if a MIDI message is sent outside of a basic channel group, the message is ignored.

To reproduce this start fluidsynth with config:

```
setbasicchannels 1 poly_omnion 1
```

Then send a "note on message" to channel 1 (index 0). It will be ignored.

The reason that I created a discussion for this issue instead of making a bug report is that it seems to be intentional behavior to support FluidSynth's concept of "basic channel groups." Was this behavior inspired by any supplemental specifications, addendums or other synths, or is it unique to FluidSynth? Is there an ideal place to document these deviations from the MIDI standard? If so I don't mind documenting that.

**Comment by derselbst** [comment]
@jjceresa Can you comment on that?

**Comment by jjceresa** [comment]
Omni On/Off, Poly/Mono in FluidSynth is documented in doc/FluidPolyMono-004.pdf.
Chapter 2 explains that fluidsynth implementation is based on the "MIDI standard specification knowledge" (MIDI Specification.pdf  1.0 version 4.2 1995).

Fluidsynth doesn't attempt to do any deviations from the MIDI standard. In particular the  "recommended start up condition"
is respected. as explained in chapter 2.3.1 "What are default 'basic channels' in FluidSynth ?".
>At synthesizer creation time (new_fluid_synth()) the synthesizer instance is set with one basic channel: basic channel 0, Omni On Poly On.
>So by default this synthesizer is a polyphonic synthesizer on all MIDI channels.


@rianhunter

>To reproduce this start fluidsynth with config:
>setbasicchannels 1 poly_omnion 1
>Then send a "note on message" to channel 1 (index 0). It will be ignored.

Sorry the reported issue cannot be reproduced. The command setbasicchannels is intended to add a basic channel group
As explained in chapter 2.7.3 this command fail if any channels overlaps any existing basic channel groups.
Example:
> setbasicchannels 1 poly_omnion 1
>fluidsynth: basic channel 1 overlaps another group
>setbasicchannels: channel   1, mode   0, nbr   1, invalid argument

**Reply by rianhunter** [reply]
Thank you for your swift reply! Very much appreciated

> > To reproduce this start fluidsynth with config:
> > setbasicchannels 1 poly_omnion 1
> > Then send a "note on message" to channel 1 (index 0). It will be ignored.
> 
> Sorry the reported issue cannot be reproduced. The command setbasicchannels is intended to add a basic channel group As explained in chapter 2.7.3 this command fail if any channels overlaps any existing basic channel groups. Example:
> 
> > setbasicchannels 1 poly_omnion 1
> > fluidsynth: basic channel 1 overlaps another group
> > setbasicchannels: channel   1, mode   0, nbr   1, invalid argument

My mistake, first you must issue the `resetbasicchannels` command, then issue the `setbasicchannels 1 poly_omnion 1`. After that is done, even though the synth is set into "Omni mode" it will ignore messages on Channel 1 (index 0). This does violate the MIDI spec. Here is an image from the document "MIDI 1.0 Detailed Specifications 4.2", page 21:

![image](https://user-images.githubusercontent.com/120658/201549521-b828719a-fb6d-482d-8591-94888e96776c.png)

You can see that when "Omni Mode" is enabled, the basic channel is ignored and messages sent to any channel are acknowledged. As further evidenced by the statement that follows this image:

> The  recognizable channel(s) refers to all MIDI channels when Omni is On, or to only the receiver's Basic Channel when Omni is Off.

**Comment by jjceresa** [comment]
There is a chapter number mistake in my previous comment:

>as explained in chapter 2.3.1 "What are default 'basic channels' in FluidSynth ?"

should be replaced by "...chapter 2.1.3..."

**Comment by jjceresa** [comment]
>Here is an image from the document "MIDI 1.0 Detailed Specifications 4.2", page 21.....

The image refer to a receiver with only one channel group starting at MIDI channel 1. But a receiver (the synthesizer) could have more than one group of channel(s). Mode message (poly, mono, omni on, omni off) are CC messages intended to a particular group that must be send to the 'channel basic' of this group. Mode message aren't intended to the synthesizer.

Reading chapters 2.1.1, 2.3, 2.4, 2.5 should help to understand all this.

**Comment by rianhunter** [comment]
> Mode message (poly, mono, omni on, omni off) are CC messages intended to a particular group that must be send to the 'channel basic' of this group.

Yes that is true for "Mode messages" but in Omni-On mode non-mode messages should be accepted from any channel. That's what the image is showing. FluidSynth does not behave this way. Instead FluidSynth regards "Omni-On" as accepting messages from `[basic_channel, basic_channel + N)` (where `N` is the value sent in the mode message, I believe this is non-standard for non "mono on" mode messages).

I see the reference to the MIDI spec page 7 in your documentation:

> Since a single instrument may function as multiple "virtual" instruments, it can thus have more than
one basic channel. Such an instrument behaves as though it is more than one receiver, and each receiver can be set to a different Basic Channel. Each of these receivers may also be set to a different mode, either by front panel controls or by Mode messages received over MIDI on each basic channel. Although not a true MIDI mode, instruments operating in this fashion are described as functioning in "Multi Mode." 

I believe this is the behavior you are attempting to implement. I have to admit that the spec is unclear how "Multi Mode" should behave when Omni is on. My interpretation would be similar to how multiple physical devices on the same MIDI bus would behave if each were in omni mode, i.e. when that "basic channel group" is in omni mode, it accepts messages from every channel. So if you were to have multiple groups in omni mode, sending a message to any channel would trigger each channel group in omni mode.

**Comment by jjceresa** [comment]
>FluidSynth does not behave this way. Instead FluidSynth regards "Omni-On" as accepting messages from [basic_channel, >basic_channel >+ N) (where N is the value sent in the mode message.

If N is > 0, FluidSynth will accept non-mode messages (i.e voices message described in MIDI specs page 9) only on these N channels starting from basic_channel.
if N is 0, FluidSynth will accept non-mode messages on all any channel from basic_channel to the next basic channel minus 1 (if any) or to synth.midi-channels minus 1.

>I believe this is non-standard for non "mono on" mode messages.

Sorry, I don't understand what you mean because I don't know what are non "mono on" mode messages.

>I believe this is the behavior you are attempting to implement.
>I have to admit that the spec is unclear how "Multi Mode" should behave.

Yes, in fluidsynth this is that behaviour that is implemented: for each fluidsynth instance, the whole set of MIDI channels (0 to synth.midi-channels minus 1) can be split in independent of MIDI channels group. Each group can bet set in a distinct mode (poly omni on, mono omni on, poly omni off, mono omni off). The first MIDI channel of a group is called 'basic channel'.

One can consider a single fluidsynth instance equivalent to what the MIDI specs page 7 refers to a "single instrument."
And a single fluidsynth "channel group" should be considered equivalent to what the MIDI specs page 7 refers to a "receiver".
All this is illustrated in doc/FluidPolyMono-004.pdf chapter 2.3, 2.4, 2.5 .

**Reply by rianhunter** [reply]
> Each group can bet set in a distinct mode (poly omni on, mono omni on, poly omni off, mono omni off). The first MIDI channel of a group is called 'basic channel'.

So going back to the original topic of this discussion, this implementation of "Omni mode" is not what the MIDI spec intends. When the instrument is in "Omni mode," midi messages are accepted regardless of what the basic channel is. Omni mode does not mean listen to all messages from `[basic_channel, basic_channel -> N)`, it means listen to messages from any channel regardless of the basic channel.

> And a single fluidsynth "channel group" should be considered equivalent to what the MIDI specs page 7 refers to a "receiver".

Sort of. The difference is that a receiver in omni mode accepts messages on all channels, not just `[basic_channel, basic_channel -> N)`. When `basic_channel != 0` or there are multiple "channel groups" FluidSynth behaves differently from the specification for "Omni mode".

**Comment by jjceresa** [comment]
Yes with this fluidsynth implementation of "Omni mode", a fluidsynth instance will listen to a channel only if this channel belong to a group of channel. Otherwise the message is ignored.

Having some MIDI messages ignored by a fluidsynth instance is useful in case of application making use of fluidsynth instance combined with hardware synthesizer. For example that allows a MIDI channel message to be played exclusively by the software synth or the hardware synthesizer.

>So going back to the original topic of this discussion, this implementation of "Omni mode" is not what the MIDI spec intends. 

As it is always possible to have all the MIDI channels range (0 to synth.midi-channels minus 1) belonging to a group of channel, this implementation is still compatible to what the MIDI spec intends.

**Comment by jjceresa** [comment]
>When basic_channel != 0 or there are multiple "channel groups" FluidSynth behaves differently from the specification for "Omni mode".

Sorry, I don't understand why you think so.

**Comment by rianhunter** [comment]
> Having some MIDI messages ignored by a fluidsynth instance is useful in case of application making use of fluidsynth instance combined with hardware synthesizer. For example that allows a MIDI channel message to be played exclusively by the software synth or the hardware synthesizer.

I agree that's useful but the MIDI standard allows you to get that behavior by turning "Omni mode" off.

>> When basic_channel != 0 or there are multiple "channel groups" FluidSynth behaves differently from the specification for "Omni mode".

> Sorry, I don't understand why you think so.

When basic_channel != 0 and the synth is in Omni mode, the synth should still accept messages from all MIDI channels. With FluidSynth this is not the case.

Thank you so much for taking the time to discuss this issue with me. I now understand better the intentions of the behavior of FluidSynth.

**Comment by jjceresa** [comment]
>I agree that's useful but the MIDI standard allows you to get that behavior by turning "Omni mode" off.

No. With MIDI standard, setting "Omni mode"  off on a "channels group" doesn't allow to make future noteon MIDI message to be ignored.
Only, Fluidsynth allows MIDI channels to be ignored using the "resetbasicchannels" command.

>When basic_channel != 0 and the synth is in Omni mode, the synth should still accept messages from all MIDI channels. With FluidSynth this is not the case.

Assuming a fluidsynth instance with 16 MIDI channels (0 to 15).
1)Assuming one intends a "channel group" starting at basic_channel 1. The channels range in this group can be 15 maximum (from 1 to 15 maximum). Fluidsynth will accept message on channels 1 to 15. Channel 0 is ignored.

2)Assuming one intends this fluidsynth instance requires accepting messages from all MIDI channels.
A new "channel group" should be added at channel basic 0. The number of member channel in this new group must be limited to 1 in order to not overlap the existing "channel group" at basic channel 1.

**Comment by rianhunter** [comment]
>> I agree that's useful but the MIDI standard allows you to get that behavior by turning "Omni mode" off.

> No. With MIDI standard, setting "Omni mode" off on a "channels group" doesn't allow to make future noteon MIDI message to be ignored.
Only, Fluidsynth allows MIDI channels to be ignored using the "resetbasicchannels" command.

There must be a misunderstanding here because I don't believe this is true. In a normal MIDI synth, you can simply turn omni mode off then change basic channel to have the synth ignore all messages that aren't sent to the basic channel.

> 1)Assuming one intends a "channel group" starting at basic_channel 1. The channels range in this group can be 15 maximum (from 1 to 15 maximum). Fluidsynth will accept message on channels 1 to 15. Channel 0 is ignored.
> 2)Assuming one intends this fluidsynth instance requires accepting messages from all MIDI channels.
A new "channel group" should be added at channel basic 0. The number of member channel in this new group must be limited to 1 in order to not overlap the existing "channel group" at basic channel 1.

This is understood but this is not how Omni mode normally behaves in MIDI.

**Comment by jjceresa** [comment]
>There must be a misunderstanding here because I don't believe this is true. In a normal MIDI synth, you can simply turn omni mode off then change basic channel to have the synth ignore all messages that aren't sent to the basic channel.

Turning omni mode off is specified in MIDI specs. But the method of changing basic channel to have the synth ignore all messages that aren't sent to the basic channel isn't specified in MIDI specs. Also the method to add/remove multiple basic channel group isn't specified. 
Consequently we cannot say that the MIDI standard allows  to control which MIDI channels must be ignored.

**Comment by jjceresa** [comment]
>1)Assuming one intends a "channel group" starting at basic_channel 1. The channels range in this group can be 15 maximum (from 1 to 15 maximum). Fluidsynth will accept message on channels 1 to 15. Channel 0 is ignored.
2)Assuming one intends this fluidsynth instance requires accepting messages from all MIDI channels.
A new "channel group" should be added at channel basic 0. The number of member channel in this new group must be limited to 1 in order to not overlap the existing "channel group" at basic channel 1.
>This is understood but this is not how Omni mode normally behaves in MIDI.

This is how the multiple "channels group" behaves in  a fluidsynth instance. This behaviour should be considered equivalent to what is described in the MIDI specs p7:
"Since a single instrument may function as multiple "virtual" instruments, it can thus have more than
one basic channel. Such an instrument behaves as though it is more than one receiver, and each receiver can be set to a different Basic Channel. Each of these receivers may also be set to a different mode, either by front panel controls or by Mode messages received over MIDI on each basic channel. Although not a true MIDI mode, instruments operating in this fashion are described as functioning in "Multi Mode."

**Reply by rianhunter** [reply]
Yes so I commented on that statement in the MIDI spec a few pages up. Even though each receiver can be set to a different mode, I don't think that necessarily implies that while they are in "Omni mode" they should only accept messages starting from their basic channel as they are supposed to do while not in omni mode. I would expect each "basic channel group" to behave as if each one were a different physical "receiver," i.e. when a basic channel group is in "Omni mode" it will accept messages from all channels regardless of their basic channel setting. If you had a physical set up of multiple devices on the same MIDI bus, this is how they would behave.

Of course the spec does not spell out exactly how "multi mode" should work, that is just the interpretation that seems most consistent to me. Either way I very much appreciate the time you put into this discussion. @derselbst @jjceresa what's the best way for me to turn this discussion into something useful in the long term for you guys?

**Comment by jjceresa** [comment]
>what's the best way for me to turn this discussion into something useful in the long term for you guys?

For me you should consider that a fluidsynth instance works as multiple "virtual" instruments (i.e multiple channels group) as described in FluidPolyMono-004.pdf. This allows to build an easy setup of one (or more) fluidsynth instance + hardware synth(s) on the same MIDI bus. This allows an easy control of the number of channels that belong to a "channel group" so that a voice message received from the MIDI bus are accepted exclusively by one "channel group" or one hardware synth. This allows an easy control change of the mode (poly/mono, omni on/off) of a single "channel group" because the new mode applies only to the channels members belonging to this group.
